### PR TITLE
New config option - holiday hours

### PR DESCRIPTION
**Pull request by edwinwills**
Unsure if this is useful for many people using this gem, however we had a recent need to set specific working hours during "holiday season" - e.g. over Christmas we'd like our working hours to be slightly different to our regular hours.

I've introduced a new config option here `holiday_hours` that's a hash of dates (in the string format `YYYY-MM-DD`, that behaves in a very similar way to `working_hours`, in that each key takes a hash of start times (e.g. it can still support multiple times within 1 day).

No worries if this isn't the direction you'd like this gem to take, just thought I'd share some code that was helpful to us - thanks for an excellent gem!

### CODE REVIEW

**Comment by jarthod** [comment]
Thanks for your contribution @edwinwills, I think this feature could be useful to others though in this state it's not really acceptable because you only changed one method which will create discrepancies (for example it won't work when going backward). In order for this to work in all cases, you'll have to check the other places where we use `config[:working_hours][time.wday]` and update them too (and add various specs for these cases). If you're not willing to spend more time on that that's fine, you can keep using this as an alpha implementation and if somebody else is interested maybe they'll take your branch and complete it.

**Comment by edwinwills** [comment]
Ah good point - I'll take a look at fixing up the other methods as well.

Whilst I do that, are specs passing on master on CI? Just want to see if it's my changes that have caused CI to have so many failing specs.

Cheers

**Comment by jarthod** [comment]
It looks like most of the failures would also be present on master and are caused by third party updates that happened since last build (bundler, etc..). We should try fixing those in another PR.

**Comment by edwinwills** [comment]
Great - I've added the config to cover:

* `advance_to_working_time`
* `advance_to_closing_time`
* `return_to_exact_working_time`
* `in_working_hours?`

And specs now cover:

* `add_seconds`
* `advance_to_working_time`
* `advance_to_closing_time`
* `in_working_hours?`

I couldn't find any specs for `return_to_exact_working_time`, so I haven't added any, but can set some up if required.

**Comment by jarthod** [comment]
About the specs on master, I just updated the travis config so they pass again so you can merge master in your branch to get green specs.

About the method updates, I still see some methods accessing `config[:working_hours]` which have not been updated, namely: `add_seconds`, `working_day?` and `working_time_between` AFAICS. Is there a good reason for that?

Thanks for you work!

**Comment by edwinwills** [comment]
`add_seconds` - I actually added a spec for this, but didn't need to add any change to the method, as it calls `advance_to_working_time`, which is already updated with the new `holiday_hours` config.

`working_day` I left as-is, as the way I expected holiday hours to work, is that you're just adjusting hours that are already present in the standard working days (so `working_day?` logic will be unchanged, as only the *hours* are changing).

`working_time_between` looks like it needs some extra work from me - will take a look and update, thanks.

**Comment by jarthod** [comment]
About `add_seconds`, yes for your spec it doesn't need changing because you only override the starting hour, and that's when we use `advance_to_working_time`, but then during the day we use `holiday_hours` to compute time, so for example if you holiday hour finish earlier, this won't be taken into account currently (or if there are two ranges and the second range start changes for example)

About `working_day` I agree for your case but nothing prevents anybody to put some exceptional days here that is usually not worked for example, and we don't want to have inconsistent results that will make people feel like the gem is not reliable. So IMO either we validate what you say at configuration time, or we make it work with every days.

Thanks again for your work and sorry for being picky about it but this is the kind of details that could made the gem untrustworthy and that's what made us loose weeks debugging others gems and finally writing our own so we really want to spare the trouble for our users. It's much easier if we adresse everything we can think of now than let people spend hours debugging why they don't get the correct result down the road.

**Comment by edwinwills** [comment]
Appreciate your thorough response! Will get back to you with fixes for the points you raised.

**Comment by edwinwills** [comment]
Hi @jarthod, somehow a whole year has gone past, and I completely forgot to update this PR, until I ran into a limitation with the `working_day?` method that you highlighted in your comment.

I've tried to fix up the remaining methods now, and added some tests to ensure that adjustments to the start and end of the day are covered.

I've also now ensured that `holiday_hours` work regardless of whether the day is an existing "working day" or not.

Thanks.

File: lib/working_hours/computation.rb [hunk 1]
```diff
@@ -246,5 +250,23 @@ def convert_to_original_format time, original
       end
     end
 
+    def holiday_hour?(time, config: nil)
+      matched_hours = []
+      config[:holiday_hours].each do |date, hours|
+        next unless time.to_date.to_s == date
+  
+        hours.each do |start_time, end_time|
+          if hour_to_seconds(time) >= start_time && hour_to_seconds(time) <= end_time
```
Comment (jarthod) [inline lib/working_hours/computation.rb:259]: I believe this `hour_to_seconds` method is the `time.seconds_since_midnight` method from ActiveSupport we use in other places in this project so you can probably use that instead to keep the code consistent.
Also you should only do this conversion one instead of N times inside the loop.
Finally, you should have the same time interval behavior we apply in `in_working_hours?` which means the end_time should be excluded.
Overall I would say you better start from the `in_working_hours?` method and modify it for your needs here as this one is quite different and less efficient.

**Review by jarthod — CHANGES_REQUESTED** [review]
Thanks for the updates @edwinwills!

I added a lot of comments to the MR and I see there are still at least one previous concern which is still not handled, not a big deal for your use maybe but to get this feature merged we'll need a bit more polishing about:

> About `add_seconds`, yes for your spec it doesn't need changing because you only override the starting hour, and that's when we use advance_to_working_time, but then during the day we use holiday_hours to compute time, so for example if you holiday hour finish earlier, this won't be taken into account currently (or if there are two ranges and the second range start changes for example)

Also it would be great if you could update the readme with instructions and examples about how to use this new feature, the new method, and the priorities between working_hours, holidays and holiday_hours.

Thanks again for your work and sorry for being a pain in the a** about it but as I said last year this is the kind of details that could made our users and us loose weeks debugging so we really want to spare the trouble. It's much easier if we adresse everything we can think of now than let people spend hours debugging why they don't get the correct result down the road

File: lib/working_hours/config.rb [hunk 2]
```diff
@@ -151,6 +177,36 @@ def validate_working_hours! week
         end
       end
 
+      def validate_holiday_hours! week
```
[the remark this answers is not in the document]
Comment (edwinwills) [reply lib/working_hours/config.rb]: Yeah good shout - switched to `days` now

File: lib/working_hours/config.rb [hunk 3]
```diff
@@ -151,6 +177,36 @@ def validate_working_hours! week
...
+        begin
+          (week.keys.select { |k| Date.parse(k) })
+        rescue => ex
+          raise InvalidConfiguration.new "Invalid day identifier(s): Must be in the format 'YYYY-MM-DD'"
+        end
```
[the remark this answers is not in the document]
Comment (edwinwills) [reply lib/working_hours/config.rb:183]: Yeah - looking back, not sure why I stuck with a string - `Date` is much nicer (and saves having to parse the keys each time). Have switched now.

File: lib/working_hours/computation.rb [hunk 4]
```diff
@@ -99,7 +99,8 @@ def advance_to_closing_time time, config: nil
         end
         # find next working range after time
         time_in_day = time.seconds_since_midnight
-        config[:working_hours][time.wday].each do |from, to|
+        time = time.beginning_of_day
```
[the remark this answers is not in the document]
Comment (edwinwills) [reply lib/working_hours/computation.rb]: Oops - removed again, thanks

File: lib/working_hours/computation.rb [hunk 5]
```diff
@@ -129,7 +130,8 @@ def return_to_exact_working_time time, config: nil
         end
         # find last working range before time
         time_in_day = time.seconds_since_midnight
-        config[:working_hours][time.wday].reverse_each do |from, to|
+        (config[:holiday_hours][time.to_date.to_s] || config[:working_hours][time.wday]).reverse_each do |from, to|
+          # round is used to suppress miliseconds hack from `end_of_day`
```
[the remark this answers is not in the document]
Comment (edwinwills) [reply lib/working_hours/computation.rb]: Removed, thanks

File: lib/working_hours/computation.rb [hunk 6]
```diff
@@ -246,5 +250,23 @@ def convert_to_original_format time, original
       end
     end
 
+    def holiday_hour?(time, config: nil)
+      matched_hours = []
+      config[:holiday_hours].each do |date, hours|
+        next unless time.to_date.to_s == date
```
[the remark this answers is not in the document]
Comment (edwinwills) [reply lib/working_hours/computation.rb]: Yeah, taking another look you're right - I don't need this method at all, as I only need to check whether the day is included in the `holiday_hours` config or not, and then `in_working_hours?` takes care of the times within the day. Removed!

File: lib/working_hours/computation.rb (hunk 1 above)
Comment (edwinwills) [reply lib/working_hours/computation.rb]: Killed the whole method now, `in_working_hours?` already takes into account any `holiday_hours` overridden for the day with `(config[:holiday_hours][time.to_date] || config[:working_hours][time.wday])`

File: lib/working_hours/config.rb [hunk 7]
```diff
@@ -151,6 +177,36 @@ def validate_working_hours! week
...
+            elsif compile_time(finish) >= 24 * 60 * 60
+              raise InvalidConfiguration.new "Invalid time: #{finish} - outside of day"
+            elsif start >= finish
+              raise InvalidConfiguration.new "Invalid range: #{start} => #{finish} - ends before it starts"
+            elsif last_time and start < last_time
+              raise InvalidConfiguration.new "Invalid range: #{start} => #{finish} - overlaps previous range"
+            end
+            last_time = finish
+          end
+        end
```
[the remark this answers is not in the document]
Comment (edwinwills) [reply lib/working_hours/config.rb:193]: Yep agree with this - have now created a `validate_hours!` method that takes a `type` keyword argument of either `:working_hours` or `:holiday_hours`, and performs the appropriate validation across the passed `dates` argument.

**Comment by edwinwills** [comment]
Appreciate all your comments, and your time to take a look through again, and totally agree with wanting to get it as simple as possible to avoid any problems in the future.

I've made the suggested changes which simplify things considerably. I've also added some notes to the Readme (but let me know if you'd like me to be a bit more verbose here).

Finally, I've added an extra spec around the `add_seconds` method - as far as I can tell it's behaving as expected (e.g. it takes into account the `holiday_hours` configuration). I now have specs covering adding seconds to the beginning, middle, and end of the day (to test that, e.g. if `holiday_hours` is set to let people have a longer lunch, the `add_seconds` takes into account a longer period of non-working time in the middle of the day.

Thanks again.

File: lib/working_hours/config.rb [hunk 8]
```diff
@@ -120,14 +146,15 @@ def compile_time time
...
 
-      def validate_working_hours! week
-        if week.empty?
-          raise InvalidConfiguration.new "No working hours given"
+      def validate_hours! dates, type:
+        case type
+        when :working_hours
+          validate_working_hours! dates
+        when :holiday_hours
+          validate_holiday_hours! dates
```
Comment (jarthod) [inline lib/working_hours/config.rb:154]: Thanks for the refactoring, this is looking a bit convoluted though to call a single method with two parameters just to then branch on this parameter. I think you should keep calling `validate_working_hours!` and `validate_holiday_hours!` from `precompiled`, and from these two methods, call `validate_hours!` after you verified the keys (so where the duplicated loop was before)

File: lib/working_hours/config.rb [hunk 9]
```diff
@@ -151,6 +178,21 @@ def validate_working_hours! week
...
+        if week.empty?
+          raise InvalidConfiguration.new "No working hours given"
+        end
+        if (invalid_keys = (week.keys - DAYS_OF_WEEK)).any?
+          raise InvalidConfiguration.new "Invalid day identifier(s): #{invalid_keys.join(', ')} - must be 3 letter symbols"
+        end
+      end
+
+      def validate_holiday_hours! days
+        if (invalid_keys = (days.keys.select{ |day| day.class.to_s != 'Date' })).any?
```
Comment (jarthod) [inline lib/working_hours/config.rb:191]: A bit more idiomatic ruby IMO:
```suggestion
        if (invalid_keys = (days.keys.reject { |day| day.is_a?(Date) })).any?
```

File: lib/working_hours/computation.rb [hunk 10]
```diff
@@ -80,7 +80,7 @@ def advance_to_working_time time, config: nil
         end
         # find first working range after time
         time_in_day = time.seconds_since_midnight
-        config[:working_hours][time.wday].each do |from, to|
+        (config[:holiday_hours][time.to_date] || config[:working_hours][time.wday]).each do |from, to|
```
Comment (jarthod) [inline lib/working_hours/computation.rb:83]: As you're doing this change in several places and some are still missing (`add_seconds`) IMO we should refactor this into a small method (something like `working_hours_for(time)`)
```suggestion
        working_hours_for(time).each do |from, to|
```
This would prevent any future refactoring from forgetting to update some places (like you did here so it's worth it ☺) and could also make testing easier if we want to add specs on this method (not require for now I would say).

**Review by jarthod — CHANGES_REQUESTED** [review]
Thanks for the quick changes and for your work on this feature!

About the refactoring part I added a couple more comments but we're almost good, it's much better :ok_hand:

About `add_seconds` unfortunately this method is still not properly supported as you did not modify it, and you managed to test some good cases which works because they depend on `advance_to_working_time` or `return_to_working_time` but you did not test the case which doesn't work yet ^^ so let me try to be more precise:

- In your "with a later starting hour", you need to start from a time inside worker hour (e.g. 11:00) and then go backward toward the opening time (10:00) by removing 2 hours for example, you'll see that instead of jumping to previous day you'll arrive at 9:00 because the `add_seconds` method does not take into account `holiday_hours` in it's code (line 43 and 59).
- Same thing in your "with an earlier ending hour" context if you start from withing working hours (e.g. 16:00) and try to add 2 hours, you'll end-up at 18:00 instead of the next day.
- In your "with an earlier starting time in the second set of hours within a day" you should probably add at least two tests: one going from morning to afternoon and one from afternoon to morning, both should be off by one hour as they'll use the 1h break of the default business instead of the custom 2h break from `holiday_hours`.

File: lib/working_hours/config.rb (hunk 8 above)
Comment (edwinwills) [reply lib/working_hours/config.rb]: Fixed

File: lib/working_hours/config.rb (hunk 9 above)
Comment (edwinwills) [reply lib/working_hours/config.rb]: Yep, much nicer, fixed

File: lib/working_hours/computation.rb (hunk 10 above)
Comment (edwinwills) [reply lib/working_hours/computation.rb]: Fixed

**Comment by edwinwills** [comment]
Sorry for the delay in update, a few things got in the way, but in the end, fixing up the PR wasn't that much effort. Hopefully this resolves the outstanding issues (including fixing up the `add_seconds` method. Thanks again for your clear feedback 🙂

**Comment by jarthod** [comment]
Thanks @edwinwills! no problem for the delay. All the fixes here looks good. Though having a last look I noticed there is still one method which wasn't updated: `#working_time_between`. I mentioned it before (but not in my last comments) so it looks like it got forgotten in the process.

**Comment by edwinwills** [comment]
Ah sorry about that - added the config to that method, and added a few new specs that were broken before this config was added. Thanks!

**Review by jarthod — APPROVED** [review]
Looks good to me now, do you want to have a look @Intrepidd?
Thanks for this contribution @edwinwills!

**Review by Intrepidd — APPROVED** [review]
Great addition, thanks !

**Comment by jarthod** [comment]
Merged and released as `1.4.0` :tada:
